File: ReverseEachColumnsMatrix/Reverse.c
/*
    Author : Pakshal Shashikant Jain 
    Date : 7/08/2024
    Program : Write a program which accept matrix and reverse the contents of each columns. 
*/

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

//Required Libraries
#include<stdio.h>
#include<stdlib.h>

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

//Function to Reverse contents of each Column of the given matrix
void ReverseColumn(int **arr,int irow,int icol)
{
    int i = 0;
    int j = 0;

    //Logic to reverse each Column of the matrix
    printf("Updated matrix is as follows : \n");
    for(i = irow - 1;i > -1;i--)
    {
        for(j = 0;j < icol;j++)
        {
            printf("%d\t",arr[i][j]);
        }
        printf("\n");
    }
}

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

//Main function
int main() 
{
    int **arr = NULL;
    int irow = 0;
    int icol = 0;
    int i = 0;
    int j = 0;

    printf("Jay Ganesh....\n");

    //Taking input from the user
    printf("Enter number of rows : ");
    scanf("%d",&irow);

    printf("Enter number of cols : ");
    scanf("%d",&icol);

    //Allocating memory for the matrix 
    arr = (int **)malloc(irow * sizeof(int *));

    for(i = 0;i < irow;i++)
    {
        arr[i] = (int *)malloc(icol *  sizeof(int));        
    }

    //Inserting elements into the matrix 
    printf("Enter elements in matrix : \n");
    for(i = 0;i < irow;i++)
    {
        printf("----Enter element in %d row : ----\n",i + 1);
        for(j = 0;j < icol;j++)
        {
            printf("Enter element no %d : ",j + 1);
            scanf("%d",&arr[i][j]);
        }
    }

    //Printing elements of the matrix 
    printf("Elements of matrix are : \n");
    for(i = 0;i < irow;i++)
    {
        for(j = 0;j < icol;j++)
        {
            printf("%d\t",arr[i][j]);
        }
        printf("\n");
    }

    //Call to ReverseColumn Function
    ReverseColumn(arr,irow,icol);

    //Deallocating memory of matrix 
    for(i = 0;i < irow;i++)
    {
        free(arr[i]);
    }

    free(arr);

   return 0;
}

File: ReverseEachColumnsMatrix/Reverse.go
/*
    Author : Pakshal Shashikant Jain 
    Date : 7/08/2024
    Program : Write a program which accept matrix and reverse the contents of each columns. 
*/

//Required Libraries
package main;
import "fmt"

//Function to Reverse each columns of the matrix 
func ReverseColumn(arr[][] int,irow int,icol int) {
	var i int = 0;
	var j int = 0;

	//Logic to reverse each columns of the matrix 
	fmt.Println("Updated matrix is as follows : ");
	for i = irow - 1;i > -1;i-- {
		for j = 0;j < icol ;j++ {
			fmt.Printf("%d\t",arr[i][j]);
		}
		fmt.Printf("\n");
	}
}

//Main Function
func main() {
	var arr[][] int = nil;
	var rows int = 0;
	var cols int = 0;
	var i int = 0;
	var j int = 0;

	fmt.Println("Jay Ganesh....");

	//Taking input from user
	fmt.Print("Enter Number of Rows : ");
	fmt.Scan(&rows);

	fmt.Print("Enter Number of Columns : ");
	fmt.Scan(&cols);

	//Allocating memory for matrix
	arr = make([][]int,rows);

	for i = 0;i < rows;i++ {
		arr[i] = make([]int,cols);
	} 
	
	//Inserting elements in matrix 
	fmt.Println("Enter elements of matrix : ");
	for i = 0;i < rows;i++ {
		fmt.Printf("---Enter elements of %d row : ---\n",(i + 1));
		for j = 0;j < cols;j++ {
			fmt.Printf("Enter element no %d : \n",(j + 1));
			fmt.Scan(&arr[i][j]);
		}
	}

	//Printing elements of matrix
	fmt.Println("Elements of matrix are : ");
	for i = 0;i < rows;i++ {
		for j = 0;j < cols;j++ {
			fmt.Printf("%d\t",arr[i][j]);
		}
		fmt.Printf("\n");
	}

	//Call to ReverseColumn function
	ReverseColumn(arr,rows,cols);
}